src/overlays/actors/ovl_Obj_Takaraya_Wall/z_obj_takaraya_wall.c

```diff
@@ -30,39 +50,450 @@ ActorInit Obj_Takaraya_Wall_InitVars = {
+ * space are closed (i.e the wall rises up).
+ */
+s32 ObjTakarayaWall_NearbySpacesClosed(s32 row, s32 column, TakarayaWallDirection direction) {
+    s32 row1;
+    s32 column1;
+    s32 adjacentSpaceRow;
+    s32 adjacentSpaceColumn;
+    s32 row2;
+    s32 column2;
+    s32 row3;
+    s32 column3;
+
+    if ((direction == TAKARAYA_WALL_DIRECTION_FRONT) || (direction == TAKARAYA_WALL_DIRECTION_BACK)) {
+        if (direction == TAKARAYA_WALL_DIRECTION_BACK) {
+            adjacentSpaceRow = row - 1;
+            row3 = row - 2;
+        } else { // TAKARAYA_WALL_DIRECTION_FRONT
+            adjacentSpaceRow = row + 1;
+            row3 = row + 2;
+        }
+        adjacentSpaceColumn = column;
+        column3 = column;
+        row1 = adjacentSpaceRow;
+        column1 = column + 1;
+        row2 = adjacentSpaceRow;
+        column2 = column - 1;
+    } else {
+        if (direction == TAKARAYA_WALL_DIRECTION_RIGHT) {
+            adjacentSpaceColumn = column - 1;
+            column3 = column - 2;
+        } else { // TAKARAYA_WALL_DIRECTION_LEFT
+            adjacentSpaceColumn = column + 1;
+            column3 = column + 2;
+        }
+        adjacentSpaceRow = row;
+        row3 = row;
+        row1 = row + 1;
+        column1 = adjacentSpaceColumn;
+        row2 = row - 1;
+        column2 = adjacentSpaceColumn;
+    }
+
+    if ((adjacentSpaceRow < 0) || (adjacentSpaceRow >= TAKARAYA_WALL_ROWS) || (adjacentSpaceColumn < 0) ||
+        (adjacentSpaceColumn >= TAKARAYA_WALL_COLUMNS) ||
+        (sTakarayaWallWallHeights[adjacentSpaceRow][adjacentSpaceColumn] == -10.0f) ||
+        ObjTakarayaWall_SpaceOpen(row1, column1) || ObjTakarayaWall_SpaceOpen(row2, column2) ||
+        ObjTakarayaWall_SpaceOpen(row3, column3)) {
+        return false;
+    }
+
+    return true;
+}
+
+void ObjTakarayaWall_BuildPath(s32 row, s32 column) {
+    s32 closedFlags = 0;
+    s32 closedCount = 0;
+    s32 randMode;
+    s32 direction;
+
+    if (ObjTakarayaWall_NearbySpacesClosed(row, column, TAKARAYA_WALL_DIRECTION_BACK)) {
+        closedCount++;
+        closedFlags |= (1 << TAKARAYA_WALL_DIRECTION_BACK);
+    }
+
+    if (ObjTakarayaWall_NearbySpacesClosed(row, column, TAKARAYA_WALL_DIRECTION_RIGHT)) {
+        closedCount++;
+        closedFlags |= (1 << TAKARAYA_WALL_DIRECTION_RIGHT);
+    }
+
+    if (ObjTakarayaWall_NearbySpacesClosed(row, column, TAKARAYA_WALL_DIRECTION_FRONT)) {
+        closedCount++;
+        closedFlags |= (1 << TAKARAYA_WALL_DIRECTION_FRONT);
+    }
+
+    if (ObjTakarayaWall_NearbySpacesClosed(row, column, TAKARAYA_WALL_DIRECTION_LEFT)) {
+        closedCount++;
+        closedFlags |= (1 << TAKARAYA_WALL_DIRECTION_LEFT);
+    }
+
+    if (closedCount == 0) {
+        sTakarayaWallPathBuilderIndex--;
+        if (!sTakarayaWallPathBuilderReachedFront) {
+            D_80ADA508--;
+        }
+    } else {
+        randMode = (s32)Rand_ZeroFloat(closedCount) % closedCount;
```

Comment:
```suggestion

        // randomly selects a new direction from the available possibilities without biases.
        randMode = (s32)Rand_ZeroFloat(closedCount) % closedCount;
```
Converted this block of code into python and computed the distribution of values, and it turns out that each direction will have an even chance of being selected. Not sure how this code achieves this but it does; I can send you the script if you want to confirm it yourself.